vdr/ambassador.go

```diff
@@ -232,13 +232,29 @@ func (n *ambassador) handleUpdateDIDDocument(transaction dag.Transaction, propos
 		WithField(core.LogFieldTransactionRef, transaction.Ref()).
 		WithField(core.LogFieldDID, proposedDIDDocument.ID).
 		Debug("Handling DID document update")
-	// Resolve latest version of DID Document
-	currentDIDDocument, _, err := n.didStore.Resolve(proposedDIDDocument.ID, &types.ResolveMetadata{AllowDeactivated: true})
-	if err != nil {
-		return fmt.Errorf("unable to update DID document: %w", err)
+
+	// Resolve version of DID Document referred to by transaction
+	var currentDIDDocument *did.Document
+	var err error
+	for _, ref := range transaction.Previous() {
```

Comment:
The `prevs` currently point to the `head` and the `doc to update`. Both refs could point to different versions of the same DID document and there is no way to keep them apart. So we cannot assume that the first ref that points to the DID doc is intended as the ref to the version of the DID doc to update.

This should probably never happen on our node™, but maybe on other implementations / (future) clustering solutions?

Reply:
Actually we can, because if the tx points to two doc updates, then those were actually conflicted and the metadata should contain refs to both tx. Therefore the resolve will find the same version if either f the prevs was used.

Reply:
I was thinking of the case where the two prevs are not conflicted. The update currently being processed is the one that will create the conflict with the current head.

For example:
tx1=DID create, [head==tx0]<=prevs
...
tx10=DID update of [tx1, head==tx9]<=prevs
tx11=DID update of [tx1, head==tx10]<=prevs

tx10 and tx11 will create a conflict, but that requires tx11 to be accepted by the VDR. If tx10 is the first ref in `prevs` of tx11, the current solution will try to update tx10 twice, while it should update tx1. Out of order processing in the VDR means that the state of a DID document is not deterministic after tx10 and tx11 have been processed.

Reply:
The VDR document store solves this. If tx10 or tx11 comes first doesn't matter. In either case a timeline is constructed and tx10 will come before tx11. Then tx11 consumes all previous states (tx10 and tx1). All nodes will mark tx11 as the latest state.

> tx10 and tx11 will create a conflict

They don't. tx11 consumes all previous source transactions.

Remember that this logic is for determining the controllers not for determining DID document state.

Reply:
edge case tracked in #1873